packages/js/src/ui/helpers/createInfiniteScroll.ts

```diff
@@ -0,0 +1,59 @@
+import { Accessor, batch, createComputed, createResource, createSignal, onCleanup, Setter } from 'solid-js';
+import { isServer } from 'solid-js/web';
+
+export function createInfiniteScroll<T>(fetcher: (page: number) => Promise<T[]>): [
+  pages: Accessor<T[]>,
+  loading: Accessor<boolean>,
+  loader: (el: Element) => void,
+  options: {
+    page: Accessor<number>;
+    setPage: Setter<number>;
+    setPages: Setter<T[]>;
+    end: Accessor<boolean>;
+    setEnd: Setter<boolean>;
+  }
+] {
+  const [pages, setPages] = createSignal<T[]>([]);
+  const [initialLoading, setInitialLoading] = createSignal(true);
+  const [page, setPage] = createSignal(0);
+  const [end, setEnd] = createSignal(false);
+
+  let add: (el: Element) => void = () => {};
+  if (!isServer) {
+    const io = new IntersectionObserver((e) => {
+      if (e.length > 0 && e[0]!.isIntersecting && !end() && !contents.loading) {
+        setPage((p) => p + 1);
+      }
+    });
+    onCleanup(() => io.disconnect());
+    add = (el: Element) => {
+      io.observe(el);
+      onCleanup(() => io.unobserve(el));
+    };
+  }
+
+  const [contents] = createResource(page, fetcher);
```

Comment:
✏ **suggestion:** please move this line higher in the code as you are referencing contents in the intersection observer code.